x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/validation/SourceDestValidator.java

```diff
@@ -0,0 +1,513 @@
+                        if (response.remoteClusterLicenseInfo().licenseInfo().getStatus() != LicenseStatus.ACTIVE) {
+                            context.addValidationError(
+                                REMOTE_CLUSTER_LICENSE_INACTIVE,
+                                true,
+                                response.remoteClusterLicenseInfo().clusterAlias()
+                            );
+                        } else {
+                            context.addValidationError(
+                                FEATURE_NOT_LICENSED_REMOTE_CLUSTER_LICENSE,
+                                true,
+                                response.remoteClusterLicenseInfo().clusterAlias(),
+                                context.getLicense(),
+                                response.remoteClusterLicenseInfo().licenseInfo().getType()
+                            );
+                        }
+                    }
+                },
+                    e -> {
+                        context.addValidationError(
+                            UNKNOWN_REMOTE_CLUSTER_LICENSE,
+                            true,
+                            context.getLicense(),
+                            remoteAliases,
+                            e.getMessage()
+                        );
+                    }
+                ), latch));
+
+            try {
+                latch.await();
+            } catch (InterruptedException e) {
+                context.addValidationError(TIMEOUT_CHECK_REMOTE_CLUSTER_LICENSE, true, context.getLicense(), remoteAliases);
+            }
+        }
+    }
+
+    static class DestinationInSourceValidation implements SourceDestValidation {
+
+        @Override
+        public boolean isDeferrable() {
+            return true;
+        }
+
+        @Override
+        public void validate(Context context) {
+            final String destIndex = context.getDest();
+
+            for (String src : context.getSource()) {
+                if (Regex.simpleMatch(src, destIndex)) {
+                    context.addValidationError(DEST_IN_SOURCE, true, destIndex, src);
+                    return;
+                }
+            }
+
+            if (context.resolvedSource.contains(destIndex)) {
+                context.addValidationError(DEST_IN_SOURCE, true, destIndex, Strings.arrayToCommaDelimitedString(context.getSource()));
+                return;
+            }
+
+            if (context.resolvedSource.contains(context.resolveDest())) {
+                context.addValidationError(
+                    DEST_IN_SOURCE,
+                    true,
+                    context.resolveDest(),
+                    Strings.collectionToCommaDelimitedString(context.resolveSource())
+                );
+                return;
+            }
+        }
+    }
+
+    static class DestinationSingleIndexValidation implements SourceDestValidation {
+
+        @Override
+        public boolean isDeferrable() {
+            return false;
```

Comment:
If callers can pass in their desired validations at call time, the idea of `isDeferrable` is no longer necessary.